### ISSUE TITLE
Error on defining constants when they're mutually exclusive

### ISSUE THREAD

**Issue by jshannon-tml** [microsoft/pyright#4871]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
I get a [reportConstantRedefinition](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportConstantRedefinition) error when defining config constants that are mutually exclusive. E.g.:

```
if os.getenv('DEBUG'):
  API_HOST = 'http://STAGING/'
else:
  API_HOST = 'http://PROD/'
```

**To Reproduce**
```
if os.getenv('DEBUG'):
  API_HOST = 'http://STAGING/'
else:
  API_HOST = 'http://PROD/'
```

**Expected behavior**
I should receive no error since I'm not redefining API_HOST.

**VS Code extension or command-line**
VS code  v1.1.301

** Context **
Asked a question about this on [SO](https://stackoverflow.com/questions/75890696/how-to-define-constants-in-python-based-on-if-else-without-pyright-warning?noredirect=1#comment133862474_75890696) and was told it was likely a pyright bug.

**Comment by erictraut** [comment]
Pyright is behaving as intended here. These are not constants from the perspective of a static type checker because their types depends on a dynamic behaviors that cannot be determined statically. A constant is a symbol whose value is assigned only once and has the same type regardless of which code paths are taken.

**Comment by erictraut** [comment]
If you want this to be interpreted as a constant by a static type checker, you'll need to tell pyright which of the two definitions you want it to assume. You can do this using `typing.TYPE_CHECKING`, for example.

```python
from typing import TYPE_CHECKING

if TYPE_CHECKING or os.getenv('DEBUG'):
  API_HOST = 'http://STAGING/'
else:
  API_HOST = 'http://PROD/'
```

**Comment by jshannon-tml** [comment]
I don't necessarily want it to be interpreted as a constant based on pyright's strict definition of a constant. But I do want to use the `UPPERCASE` convention to be consistent with the colloquial definition and use of constants within Python.

I get what you're saying about "depends on a dynamic behaviors that cannot be determined statically". In fact, I tried to pre-define the type:

```
API_HOST: str

if os.getenv('DEBUG'):
  API_HOST = 'http://STAGING/'
else:
  API_HOST = 'http://PROD/'
```

Which I feel should address the type checkers possible confusion. Specifically, I'm defining API_HOST as a string and while I _could_ assign non-strings to API_HOST later, I've been explicit about what its type _should_ be and the type checker can also throw an error if I assign a non-string. In fact, pyright accepts:

```
API_HOST = 'http://...' if os.getenv('DEBUG') else 123
```
and determines API_HOST to be a union of two literals (as expected). But that's fundamentally pretty similar to what I initially tried to do.

**Comment by erictraut** [comment]
Most programming languages provide a keyword for defining constants. Unfortunately, Python has no such keyword. The convention is to use all upper case for constants, and pyright uses this convention when enforcing its `reportConstantRedefinition` rule. This rule is a strict-mode check, so it is somewhat more opinionated than the "basic" type checking rules. You're welcome to disable it if you don't agree with the heuristics it uses.
